Xenial / Systemd / Offline support

lazypower opened the pull request:
This branch introduces some very different behavior from the install mechanism, but overall simplifies delivery of the charm.
- Offlineable resources via juju_resources
- Support for SystemD based hosts (verified on Xenial)
- Cross arch support via juju_resources

Stems from #8

Commit:
Adds xenial and trusty to metadata

This charm layer should support trusty/xenial, reflect this in metadata

Commit:
Supporting offline resource files

These bits add a make target for `fetch_resources`

- Parse the resources.dist file in $CHARM_ROOT
- Fetch and sha1sum the resources parsed out of resources.dist
- Print feedback about sum verification, all files isolated in `$CHARM_ROOT/resources`

Commit:
Remove older install code

Commit:
Resource Delivery Pattern

- Added resource delivery pattern
- Verified on sale of up to 3 nodes
- removes some extra output from scripts/download_resources

Commit:
Final tweaking for xenial/systemd hosts

This wraps the Defaults, Systemd, and final location of the bins (to /usr/bin/%p) 

Passes scale test to 5 nodes with a healthy cluster report

Commit:
Throw in some Documentation updates

Touching up the lacking README

Commit:
Readme Touchups

Additional readme information relating to the binaries. 

This should wrap up #8

Commit:
bump etcd version to latest stable

nelljerram commented on README.md:
````diff
@@ -23,7 +23,51 @@ juju add-unit -n 2 etcd
+### Supply your own binaries
+
+```shell
+juju upgrade-charm etcd --resource etcd=./path/to/etcd --reosurce etcdtcl=./path/to/etcdctl
````
Typo 'reosurce'

nelljerram commented on README.md:
```diff
@@ -23,7 +23,51 @@ juju add-unit -n 2 etcd
+### If you do not have the binaries
+
+If you are simply deploying etcd, and the charm has halted demanding resources by
+telling you to consult the READJE, you can use a script contained in the charm
```
Typo 'READJE'

nelljerram commented on README.md:
```diff
@@ -23,7 +23,51 @@ juju add-unit -n 2 etcd
+
+This charm supports resources, which means you can supply your own release of
+etcd and etcdctl to this charm. Which by nature makes it highly deployable on
+multiple architectures.
```
As also would just referencing a package name, surely?

lazypower replied:
No. Juju resources are decoupled from package management.  This provides a streamlined way to provide upstream binaries to the charm. It has a fallback mechanism on xenial hosts to reference archive if no resources are supplied.

This wont work on trusty, as there is no etcd package in archive to support the request. This created the install tree that has 3 paths.

Upstart + resources
Systemd + resources
Systemd + archive packages

Its not implicitly called out this way, as its preferrable to use the resources that will ship with the charm. As a charm maintainer, we can rev the resources on a development channel of the charm and run smoke tests to verify everything is good, and then upgrade the resource payload on the stable channel of the charm to deliver the latest resource. 

This enables offline deployments as well as supporting upstreams best efforts to have patch releases pushed quickly to consumers of the charm.

nelljerram replied:
OK - thanks for the explanation.

nelljerram commented on README.md:
```diff
@@ -23,7 +23,51 @@ juju add-unit -n 2 etcd
 
+## Usage with your own binaries
+
+This charm supports resources, which means you can supply your own release of
```
'can' or 'have to'?  If it's optional, why would anyone ever want to do that, and why would the charm allow them to?

lazypower replied:
'can'.  Its optional as you may be running on PPC64EL architecture and need to supply a non AMD64 bin for the workload to function, or perhaps the charm store is still deploying v 2.3.3 and v3.0.0 just hit GA and you wish to smoke test it in your own environment. This prevents consumers from having to rev packaging/the-charm to do these operations.

nelljerram commented on README.md:
```diff
@@ -23,7 +23,51 @@ juju add-unit -n 2 etcd
 Its recommended to run an odd number of machines as it has greater redundancy than even number (ie. 4, you can lose 1 before quorum is lost, where as 5, you can 2).
 
+# The charm told me to see the README
 
```
So here's what I hope the situation is, after all these changes.
- If you're deploying on Xenial or later, the charm installs the etcd package from the Ubuntu archive, for any supported architecture.
- Otherwise, if you're deploying on amd64, the charm grabs and uses a binary tarball from the CoreOS github account.
- Otherwise - i.e. both old distro and weird architecture - you have to provide the resources yourself.

If that's right, here's what I think the README should say:

You've been directed here because you're deploying this etcd charm onto a pre-Xenial and non-amd64-based Ubuntu platform, and we don't have etcd and etcdctl binaries for that platform.  You will need to obtain such binaries somehow yourself (e.g. by downloading and building from source), then tell Juju about those binaries as detailed below.

lazypower replied:
That sounds far better than how I was trying to articulate it. Thanks :)

nelljerram commented on Makefile:
```diff
@@ -0,0 +1,28 @@
+	tox
+
+.PHONY: fetch_resources
+fetch_resources:
```
I don't get why this is a Makefile target.  IIUC correctly, the charm itself should automatically determine if it needs to download resources, and if so, do that at the appropriate time.  It shouldn't be presented as a user choice.

lazypower replied:
This is more for my sake, while I'm revving the charm testing different bins and preparing for a publish this script keeps me honest. I've also leveraged it in the README for users wishing to explore the functionality. (it wasn't originally worded this way)

its safe to ignore from a user perspective, but important from a release management perspective.

nelljerram replied:
Understood - thanks.

nelljerram commented on scripts/download_resources.sh:
```diff
@@ -0,0 +1,24 @@
+
+set -e
+
+#!/bin/bash
```
Second hashbang line here.

nelljerram commented on reactive/etcd.py:
```diff
@@ -180,10 +180,12 @@ def configure_etcd():
         # render systemd
-        templating.render('systemd', 'files/systemd/etcd.service',
+        templating.render('systemd', '/etc/systemd/system/etcd.service',
                           cluster_data, owner='root', group='root')  # noqa
```
This looks wrong to me, because on Xenial there should already be an etcd.service file installed by the package (in /usr/lib/systemd/system), and the 'systemctl enable etcd' call just below will symlink that into /etc/systemd/system.  So the charm shouldn't need to generate an independent copy of the service file.

lazypower replied:
Ah thats true. We only need to install this file if we've done this via resources. This did cause me some headache and I think I missed a patch I had made remotely to fix this. Good catch

nelljerram commented on templates/defaults:
```diff
@@ -4,7 +4,7 @@ ETCD_ADVERTISE_CLIENT_URLS="http://{{ private_address }}:{{ port }}"
 ETCD_LISTEN_PEER_URLS="http://{{ private_address }}:{{ management_port }}"
 ETCD_INITIAL_ADVERTISE_PEER_URLS="http://{{ private_address }}:{{ management_port }}"
-ETCD_ADVERTISE_PEER_URLS="http://172.31.20.186:2380"
+ETCD_ADVERTISE_PEER_URLS="http://{{ private_address}}:{{ management_port }}"
```
Does it matter that there isn't a trailing space after private_address?

nelljerram commented on README.md:
```diff
@@ -1,11 +1,30 @@
-> This charm was built from layers! To see the developer docs see `hacking.md` in
-the charm root.
+Etcd is a highly available distributed key value store that provides a reliable way to store data across a cluster of machines. Etcd gracefully handles master elections during network partitions and will tolerate machine failure, including the master.
+
```
I personally prefer text wrapped <=79 columns (e.g. to read more easily in gitk).

nelljerram commented on README.md:
````diff
@@ -1,11 +1,30 @@
+```shell
+juju deploy etcd
+```
+Add and capacity with:
````
Should be 'And add', I think.

nelljerram commented on README.md:
````diff
@@ -1,11 +1,30 @@
+juju add-unit -n 2 etcd
+```
+
+Its recommended to run an odd number of machines as it has greater redundancy than even number (ie. 4, you can lose 1 before quorum is lost, where as 5, you can 2).
````
Should be It's
s/even/an even/
s/4/with 4/
s/where as 5, you can/whereas with 5, you can lose/

Commit:
Review Comments

implementing reivew comments from @nieljerram

mbruzek commented on reactive/etcd.py:
```diff
@@ -97,13 +105,53 @@ def update_cluster_string():
+            host.adduser('etcd')
+            host.add_user_to_group('etcd', 'etcd')
+
+        os.makedirs('/var/lib/etcd/', exist_ok=True)
```
exist_ok is only valid in 3.2 and up, making this a 2.7 incompatible change. I see you changed the shebang to python3 so this is fine, but it takes only one more line to make it compatible with 2.7 (here)

``` python
if not os.isdir('/var/lib/etcd')
    os.makedirs('/var/lib/etcd)
```

lazypower replied:
Thats as fair point, but I'd rather let python 2.7 silently die off. Charms  have migrated to python3 since we started the layer approach. I'd like to keep pushing the envelope of making the transition.

Commit:
Catch errors for 1.25 users

Resources landed in 2.0+ clients, this should give users of 1.25 better error messaging and prompt them to upgrade.

mbruzek commented on reactive/etcd.py:
```diff
@@ -97,13 +105,57 @@ def update_cluster_string():
+        etcd_path = resource_get('etcd')
+        etcdctl_path = resource_get('etcdctl')
+    except NotImplementedError:
+        status_set('blocked', 'This charm requires the resource feature available in juju 2+')  # noqa
```
This line will give juju 1.x users a good message to get on the 2 train!

mbruzek commented on templates/systemd:
```diff
@@ -0,0 +1,22 @@
+Type=notify
+User=etcd
+PermissionsStartOnly=true
+ExecStart=/usr/bin/etcd $DAEMON_ARGS
```
I don't systemd very well yet. Should you not use the Jinja template variables in this file as well? How are the default values getting in this systemd file?

lazypower replied:
line 11 points it at /etc/default/etcd, which is the rendered form of the defaults template
